Mathlib/Analysis/Normed/Field/Ultra.lean

```diff
@@ -0,0 +1,146 @@
+/-
+Copyright (c) 2024 Yakov Pechersky. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yakov Pechersky
+-/
+import Mathlib.Analysis.Normed.Field.Basic
+import Mathlib.Analysis.Normed.Group.Ultra
+
+/-!
+## Sufficient contidition to have an ultrametric norm on a field
+
+This file provides ways of constructing an instance of `IsUltrametricDist` based on
+facts about the existing norm.
+
+## Main results
+
+* `isUltrametricDist_of_forall_norm_natCast_le_one`: a norm in a field is ultrametric if the norm
+  of the image of a natural is less than or equal to one
+
+## Implementation details
+
+The proof relies on a bounded-from-above argument, so the imports need to include `rpow`
+
+## Tags
+
+ultrametric, nonarchimedean
+-/
+open Metric IsUltrametricDist NNReal
+
+namespace IsUltrametricDist
+
+section sufficient
+
+variable {R F : Type*} [NormedDivisionRing R] [NormedField F]
+
+lemma isUltrametricDist_of_forall_norm_add_one_le_max_norm_one
+    (h : ∀ x : R, ‖x + 1‖ ≤ max ‖x‖ 1) : IsUltrametricDist R := by
+  apply isUltrametricDist_of_forall_norm_add_le_max_norm
+  intro x y
+  rcases eq_or_ne y 0 with rfl|hy
+  · simp
+  rw [← div_le_div_right (c := ‖y‖) (by simpa using hy), ← norm_div, add_div, div_self hy,
+      ← max_div_div_right (norm_nonneg _), div_self (by simp [hy]), ← norm_div]
+  exact h _
+
+lemma isUltrametricDist_of_forall_norm_sub_one_of_norm_le_one
+    (h : ∀ x : R, ‖x‖ ≤ 1 → ‖x - 1‖ ≤ 1) : IsUltrametricDist R := by
+  have : ∀ x : R, ‖x‖ ≤ 1 → ‖x + 1‖ ≤ 1 := by
+    intro x hx
+    specialize h (-x) (by simpa using hx)
+    rwa [← neg_add', norm_neg] at h
+  apply isUltrametricDist_of_forall_norm_add_one_le_max_norm_one
+  intro x
+  cases le_or_lt ‖x‖ 1 with
+  | inl h => simpa [h] using this _ h
+  | inr h =>
+    suffices ‖x + 1‖ ≤ ‖x‖ by simp [this]
+    rw [← div_le_div_right (c := ‖x‖) (h.trans' zero_lt_one), div_self (h.trans' zero_lt_one).ne',
+        ← norm_div, add_div, div_self (by simpa using (h.trans' zero_lt_one)), add_comm]
+    apply this
+    simp [inv_le_one_iff₀, h.le]
```

Comment:
The structure of this proof is a little strange since you only use the main hypothesis in the initial `have` block. Maybe worth splitting off a new lemma whose hypothesis is `∀ x : R, ‖x‖ ≤ 1 → ‖x + 1‖ ≤ 1`?

Golfs of the proof:
```suggestion
lemma isUltrametricDist_of_forall_norm_sub_one_of_norm_le_one
    (h : ∀ x : R, ‖x‖ ≤ 1 → ‖x - 1‖ ≤ 1) : IsUltrametricDist R := by
  have {x : R} (hx : ‖x‖ ≤ 1) : ‖x + 1‖ ≤ 1 := by
    simpa only [← neg_add', norm_neg] using h (-x) (norm_neg x ▸ hx)
  refine isUltrametricDist_of_forall_norm_add_one_le_max_norm_one fun x ↦ ?_
  rcases le_or_lt ‖x‖ 1 with h | h
  · exact (this h).trans (le_max_right _ _)
  · suffices ‖x + 1‖ ≤ ‖x‖ from this.trans (le_max_left _ _)
    have p : 0 < ‖x‖ := h.trans' zero_lt_one
    rw [← div_le_div_right p, div_self p.ne', ← norm_div, add_div, div_self (norm_pos_iff.mp p),
      add_comm]
    exact this (by simpa only [one_div, norm_inv] using inv_le_one_of_one_le₀ h.le)
```